tests/test_affinity_request_routing.py

```diff
@@ -0,0 +1,178 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
```

Comment:
I suppose that checking from log is bad idea.
We can just test `get_best_node` and test the same thing.

Reply:
Using logs to check correct request routing is the most robust approach.

Sure, you can test `get_best_node`, but what if we forget to call it in one of the cache methods by mistake? The tests will pass, but the code is broken.

With the current approach we check every single method that should be partition-aware.

Reply:
There is another variant -- monkey patch get_best_node to function with side effects and check these effects.
This solution is not robust at all. It's slow, buggy and ugly.

Reply:
Changing the implementation in order to test it is called "test-induced design damage". With the current approach we can refactor the implementation entirely - the test will not require any changes. 

> slow
How slow? Did you measure it?

> buggy
This testing approach is used by other thin clients as well, not a single bug in more than a year.

> ugly
Please elaborate.

Reply:
> How slow? Did you measure it?
Tests on master more than 4 times faster than current approach.
> ugly
This is not a pythonic way. Pythonic way is monkey patching (so called spy)
> Changing the implementation in order to test it is called "test-induced design damage"
You can change method in runtime easily, please read about monkey patching. This techinque give ability
to change behaviour of object in runtime without touching main code at all.

Reply:
BTW, current tests is broken for SSL :) And, to be honest, it will be nice to fix them.

Reply:
Ok, I don't insist, this is not a big problem. Current python thin client is a much more problem, it's broken by design.

Reply:
> Tests on master more than 4 times faster than current approach.

Are you serious? :) master has 2x less tests, this comparison does not tell us much.

Anyway, I don't insist as well. I've explained the reasoning, now it is up to @isapego to decide.

Reply:
Tests with/with-password/without ssl is faster then tests without ssl with current approach. May be grepping in logs can be optimized, may be it's because start-stop of servers.

Reply:
Ivan, I believe that tests now are slower just because there are more tests and now we start 3 to 4 nodes instead of one node before. Yeah, tests before were faster but they covered less cases. I don't think that grepping the log really takes a lot of time.

Reply:
Yes, I changed my opinion right now. It seems that indeed, just starting-stopping multiple nodes is the main reason of slow down.

I also created PR to PR in order to fix tests with ssl and travis.